README.md
ImputeRobust: Multiple Imputation with GAMLSS
=============================================

The `ImputeRobust` package adds to [https://cran.r-project.org/web/packages/mice/index.html](MICE) an imputation method based on generalized additive models for location, scale, and shape introduced by de Jong (2012), de Jong, van Buuren and Spiess (2016). It has been tested mostly with continous variables, count, and binary data.

Installation
------------

The latest version can be installed from GitHub as follows:

``` r
install.packages("devtools")
devtools::install_git(url = "https://github.com/dsalfran/ImputeRobust", 
                      branch = "master")
```

Main functions
--------------

The main function in the `ImputeRobust` package is `mice.impute.gamlss()` which model the response distribution with a Normal distribution. All other functions are variants of the main function with a different fixed response distribution.

| Function Name              | Response distribution  |
|----------------------------|------------------------|
| `mice.impute.gamlss()`     | Normal                 |
| `mice.impute.gamlssBI()`   | Binomial               |
| `mice.impute.gamlssJSU()`  | Johnson's SU           |
| `mice.impute.gamlssPO()`   | Poisson                |
| `mice.impute.gamlssTF()`   | t-distribution         |
| `mice.impute.gamlssZIBI()` | Zero inflated binomial |
| `mice.impute.gamlssZIP()`  | Zero inflated Poisson  |

Examples
--------

The package is intended to be used with mice, adding new methods based on GAMLSS. For example, let's say we have a data sets with missing values:

``` r
# First lines of the data set
head(sample.data)
#>           X.1 X.2       X.3 X.4         y
#> 1 -0.89662664  NA        NA  NA  6.943309
#> 2 -1.75990198  NA        NA   1  5.272678
#> 3  0.60351069   4 -2.330010   0  3.669100
#> 4 -0.05911172   2 -1.010365   0 -2.995331
#> 5  0.95021728   4        NA   1  5.736138
#> 6  1.61066030  NA        NA   1 18.415576
```

``` r
# Missing pattern
md.pattern(sample.data)
#>     X.1 y X.4 X.2 X.3    
#> 125   1 1   1   1   1   0
#>  39   1 1   1   1   0   1
#>  64   1 1   1   0   0   2
#>  22   1 1   0   0   0   3
#>       0 0  22  86 125 233
```

``` r
# Create the imputed data sets
imputed.sets <- mice(sample.data,
                     method = c("gamlss", "gamlssPO",
                                "gamlssJSU", "gamlssBI", "gamlss"),
                     visitSequence = "monotone",
                     maxit = 1, seed = 97123)
```

In the following plot the red circles represent the imputed values and the blue circles the observed data. ![](sample.png)

References
----------

de Jong, R., van Buuren, S. & Spiess, M. (2016) Multiple Imputation of Predictor Variables Using Generalized Additive Models. Communications in Statistics -- Simulation and Computation, 45(3), 968--985.

de Jong, Roel. (2012). “Robust Multiple Imputation.” Universität Hamburg. .

Rigby, R. A., and Stasinopoulos, D. M. (2005). Generalized Additive Models for Location, Scale and Shape. Journal of the Royal Statistical Society: Series C (Applied Statistics) 54 (3): 507–54.


NEWS
ImputeRobust v1.3-1 (Release date: 2018-11-25)
==============

* Disables logs.

ImputeRobust v1.3 (Release date: 2018-06-21)
==============

* Fixed mismatch of indexes in 'mice.impute.gamlss'.

ImputeRobust v1.2 (Release date: 2017-08-18)
==============

* Is is possible for the user to modify the imputation models
* Increased stability of the internal fitting algorithms

ImputeRobust v1.1-2 (Release date: 2017-04-20)
==============

* Fixed error in 'ImpGamlsFit' that prevented the Imputation algorithm
  to run in Windows

ImputeRobust v1.1-1 (Release date: 2017-02-24)
==============

* Moved formal arguments in 'fit.gamlss' to the inner fitting function
  due to an error where formal arguments were being matched multiple
  times.



R/sample.R
#' Sample data set with a monotone missing pattern
#'
#' A simple data set with monotone missing pattern
#'
#' Sample data set with four predictors and a dependent variable. A
#' missing monotone pattern was generated in three predictors to
#' illustrate the gamlss imputation method.
#'
#' For the data generation process a parameter beta equal to
#' \code{c(1.3, .8, 1.5, 2.5)} and a predictor matrix \code{X <-
#' cbind(X.1, X.2, X.3, X.4)} are defined. Then, the sample data set
#' is created with the model \code{y ~ X.1 + X.2 + X.3 + X.4}.
#'
#' @name sample.data
#' @docType data
#' @format A data frame with 200 rows on the following 5 variables
#'   \describe{
#' \item{X.1}{Numeric variable from a Normal distribution}
#' \item{X.2}{Count data from a Poisson distribution}
#' \item{X.3}{Numeric variable from a Normal distribution}
#' \item{X.4}{Binary variable from a Binomial distribution}
#' \item{y}{Response variable}
#' }
#' @keywords datasets
#'
#'
#' @examples
#'
#' head(sample.data)
#' 
NULL


R/ImputeRobust.R
#' Multiple Imputation with Generalized Additive Models for Location,
#' Scale, and Shape
#'
#' De Jong (2012), De Jong, van Buuren and Spiess (2016) introduced a
#' new imputation method based on generalized additive models for
#' location, scale, and shape (Rigby and Stasinopoulos, 2005), which
#' is a class of univariate regression models, where the assumption of
#' an exponential family is relaxed and replaced by a general
#' distribution family. This allows the a more flexible modelling than
#' standard parametric imputation models of not only the location
#' (e.g. the mean), but also the scale (e.g. variance), and the
#' shape (e.g., skewness and kurtosis) of the conditional distribution
#' of the dependent variable given all other variables.
#'
#' @author Daniel Salfran \email{daniel.salfran@uni-hamburg.de}
#' @author Martin Spiess \email{martin.spiess@uni-hamburg.de}
#'
#' @references de Jong, R., van Buuren, S. & Spiess, M. (2016)
#'   Multiple Imputation of Predictor Variables Using Generalized
#'   Additive Models. Communications in Statistics -- Simulation and
#'   Computation, 45(3), 968--985.
#'
#' @references de Jong, Roel. (2012). “Robust Multiple Imputation.” Universität
#' Hamburg. \url{http://ediss.sub.uni-hamburg.de/volltexte/2012/5971/}.
#'
#' @references Rigby, R. A., and Stasinopoulos,
#'   D. M. (2005). Generalized Additive Models for Location, Scale and
#'   Shape. Journal of the Royal Statistical Society: Series C
#'   (Applied Statistics) 54 (3): 507–54.
#'
"_PACKAGE"


R/model_creator.R
#' Model creator
#'
#' @description This is a helper function to be used within the gamlss
#'   fitting procedure. It creates automatically a formula object for
#'   the variables named a given data frame. The dependent variable is
#'   the one in the first column and the rest are treated as
#'   independent.
#'
#' @param data Data frame that will provide the named variables.
#' @param gam.model List of mode parameter, containing the "type" with
#'   c("linear", "cs", "pb") as available choices and "par", an
#'   optional list parameter if the model is not linear.
#' @param lin.terms Specify which predictors should be included
#'   linearly. For example, binary variables can be added directly as
#'   an additive term instead of defining a spline.
#'
#' @return Returns a formula object.
ModelCreator <- function(data, gam.model, lin.terms = NULL){
  if (class(data) != "data.frame") {
    stop("'data' must be a data frame")
  }

  type <- gam.model$type
  par <- gam.model$par
  
  dependent <- names(data)[1]
  factors <- names(data)[-1]

  if (!is.null(lin.terms)) {
    idx <- match(lin.terms, factors)
    v <- vector(mode = "logical", length = length(factors))
    v[idx] = TRUE
    factors <- factors[!v]
  }

  if (type == "response") {
    response <- factors[length(factors)]
    formula <- as.formula(paste(dependent, "~", response))
    return(formula)
  }

  if (type == "p-response") {
    response <- factors[length(factors)]
    formula <- as.formula(paste(dependent, "~",
                                paste("pb(", response, ")")))
    return(formula)
  }

  if (type == "linear" && is.null(lin.terms)) {
    # Define a linear model
    formula <- as.formula(paste(paste(dependent, " ~ ", sep = ""),
                                paste(factors, collapse = " + ")))
  } else if (type == "linear" && !is.null(lin.terms)) {
    formula <- as.formula(paste(paste(dependent, " ~ ", sep = ""),
                                paste(factors, collapse = " + "), "+",
                                paste(lin.terms, collapse = "+")))
  } else if (type == "cs") {
    # Define a cubic spline model
    if (is.null(par)) {
      df = 1
    } else {
      df = par
    }
    if (is.null(lin.terms)) {
      formula <- as.formula(paste(
        paste(dependent, " ~ ", sep = ""),
        paste("cs(", factors, " , df = ", df, ")", sep = "",
              collapse = "+")))
    } else {
      formula <- as.formula(paste(
        paste(dependent, " ~ ", sep = ""),
        paste("cs(", factors, " , df = ", df, ")", sep = "",
              collapse = "+"), "+", paste(lin.terms, collapse = "+")))
    }
  } else if (type == "pb") {
    # Define a P-spline model
    if (is.null(par)) {
      control = "control = pb.control(degree = 2, order = 2)"
    } else {
      control = paste("control = pb.control(degree = ", par$degree,
                      ", order = ", par$order, ")", sep = "")
    }
    if (is.null(lin.terms)) {
      formula <- as.formula(paste(
        paste(dependent, " ~ ", sep = ""),
        paste("pb(", factors, " ,", control, ")", sep = "",
              collapse = "+")))
    } else {
      formula <- as.formula(paste(
        paste(dependent, " ~ ", sep = ""),
        paste("pb(", factors, " ,", control, ")", sep = "",
              collapse = "+"), "+", paste(lin.terms, collapse = "+")))
    }
  } else {
    stop("Wrong choice of model")
  }

  return(formula)
}


R/imp_gamlss_fit.R
#' GAMLSS imputation fit
#'
#' @description
#' This function takes a data set to fit a gamlss model and
#' another to predict the expected parameters values. It returns a
#' function that will generate a vector of random observations for the
#' predicted parameters. The amount of random observations is the
#' number of units on the dataset used to get such predictions.
#'
#' @param data Completely observed data frame to be used to fit a
#'   gamlss model estimate.
#' @param new.data Data frame used to predict the parameter values for
#'   some given right side x-values on the gamlss model.
#' @param family Family to be used for the response variable on the
#'   GAMLSS estimation.
#' @param n.ind.par Number of individual parameters to be
#'   fitted. Currently it only allows one or two because of stability
#'   issues for more parameters.
#' @param gam.mod list with the parameters of the GAMLSS imputation
#'   model.
#' @param mod.planb list with the parameters of the alternative GAMLSS
#'   imputation model.
#' @param n.par.planb number of individual parameters in the
#'   alternative model.
#' @param lin.terms Character vector specifying which (if any)
#'   predictor variables should enter the model linearly.
#' @param n.cyc number of cycles of the gamlss algorithm
#' @param bf.cyc number of cycles in the backfitting algorithm
#' @param cyc number of cycles of the fitting algorithm
#' @param forceNormal Flag that if set to 'TRUE' will use a normal
#'   family for the gamlss estimation as a last resource.
#' @param trace whether to print at each iteration (TRUE) or not
#'   (FALSE)
#' @param ... extra arguments for the control of the gamlss fitting
#'   function
#'
#' @return Returns a method to generate random samples for the fitted
#'   gamlss model using "new.data" as covariates.
ImpGamlssFit <- function(data, new.data, family, n.ind.par, gam.mod,
                         mod.planb = list(type = "pb", par = list(degree = 1, order = 1)),
                         n.par.planb = n.ind.par, lin.terms = NULL,
                         n.cyc = 5, bf.cyc = 5, cyc = 5,
                         forceNormal = FALSE, trace = FALSE, ...) {


  # Really ugly way of specifying the formulas for the gamlss fitting
  # function
  mu.f1 <- ModelCreator(data, gam.mod, lin.terms)
  sigma.f1 <- switch(n.ind.par, ~1, mu.f1, mu.f1, mu.f1)
  nu.f1 <- switch(n.ind.par, ~1, ~1, mu.f1, mu.f1)
  tau.f1 <- switch(n.ind.par, ~1, ~1, ~1, mu.f1)

  #Plan B
  mu.planb <- ModelCreator(data, mod.planb, lin.terms)
  sigma.planb <- switch(n.par.planb, ~1, mu.planb, mu.planb, mu.planb)
  nu.planb <- switch(n.par.planb, ~1, ~1, mu.planb, mu.planb)
  tau.planb <- switch(n.par.planb, ~1, ~1, ~1, mu.planb)
  if (forceNormal) {family.planb <- NO} else {family.planb <- family}

  # Disable warnings from GAMLSS, the imputing algorithm is shown to work
  # even in the GAMLSS fitting algorithm doesn't converge
  oldw <- getOption("warn")

  tryCatch({
    # Fit gamlss model with given formula for all four moments
    # (ignored if not needed) and the given distribution family
    fit <- tryCatch(
      {options(warn = -1)
      gamlss(formula = mu.f1,
             sigma.formula = sigma.f1,
             nu.formula = nu.f1,
             tau.formula = tau.f1,
             family = family,
             data = data,
             control = gamlss.control(n.cyc = n.cyc, trace = trace, ...),
             i.control = glim.control(bf.cyc = bf.cyc, cyc = cyc, ...))},
      error = function(e) {
        tryCatch(
          ## cat("PlanB\n")
          {options(warn = -1)
          gamlss(formula = mu.f1,
                 sigma.formula = sigma.f1,
                 nu.formula = nu.f1,
                 tau.formula = tau.f1,
                 family = family,
                 data = data,
                 control = gamlss.control(n.cyc = min(3, n.cyc), trace = trace , ...),
                 i.control = glim.control(bf.cyc = min(4, bf.cyc), cyc = min(3, cyc), ...))},
          error = function(e) {
            tryCatch(
              ## cat("PlanC\n")
              {options(warn = -1)
              gamlss(formula = mu.f1,
                     sigma.formula = sigma.f1,
                     nu.formula = nu.f1,
                     tau.formula = tau.f1,
                     family = family,
                     data = data,
                     control = gamlss.control(n.cyc = min(3, n.cyc), trace = trace , ...),
                     i.control = glim.control(bf.cyc = min(3, bf.cyc), cyc = min(2, cyc), ...))},
              error = function(e) {
                ## cat("PlanD\n")
                options(warn = -1);
                gamlss(formula = mu.planb,
                       sigma.formula = sigma.planb,
                       nu.formula = nu.planb,
                       tau.formula = tau.planb,
                       family = family.planb,
                       data = data,
                       control = gamlss.control(n.cyc = min(3, n.cyc), trace = trace , ...),
                       i.control = glim.control(bf.cyc = min(3, bf.cyc), cyc = min(2, cyc), ...))
              }
            )
          }
        )
      }
    )

    options(warn = oldw)

    # Predict the parameters values for the units with missings
    capture.output(predictions <- predictAll(fit, new.data, type="response",
                                             data = data),
                   file = NULL)

    # Return wrapper that will call the generation function
    # corresponding to the distribution family of the gamlss
    # model
    function(...) {
      r <- paste("r", fit$family[1], sep = "")

      predictions$y <- NULL
      do.call(r,
              args = c(predictions,
                       n = length(predictions$mu)
                       )
              )
    }
  },
  error = function(e) {
    function(...) {do.call(rep, args = list(NA, nrow(new.data)))}
  })

}


R/imp_gamlss_bootstrap.R
#' GAMLSS bootstrap method
#'
#' @description Creates a random generation function for the missing
#'   values with bootstrap sample from the fitted GAMLSS model for the
#'   completely observed data.
#'
#' @param incomplete.data Data frame with missings on one variable.
#' @param R Boolean matrix with the response indicator.
#' @param fit Random sample generator method.
#' @param ... extra arguments for the control of the gamlss fitting
#'   function
#'
#' @return Returns a imputation sample generator.
ImpGamlssBootstrap <- function(incomplete.data, fit, R, ...) {
  ## Imputation using the bootstrap predictive distribution

  available.cases <- subset(incomplete.data, R)
  ## Fit initial gamlss model with completely observed data.
  master.predict <- fit(available.cases, available.cases, ...)

  tryCatch(
  {
    function(...) {
      bootstrap.sample <- available.cases
      repeat {
        ## Replace observed part of the available cases with a
        ## random sample of the same distribution
        bootstrap.sample[,1] <- master.predict(...)
        ## Create random sample generator with
        ## bootstrap.sample as completely observed set and the
        ## observed part of the variables with missings as
        ## predictors.
        bootstrap.predict <- fit(bootstrap.sample,
                                 subset(incomplete.data, !R), ...)
        break
      }
      bootstrap.predict(...)
    }
  }, error = function(e) {
    function(model) {
      do.call(rep,
              args=list(NA, nrow(subset(incomplete.data, !R))))}
  }
  )
}


R/mice_impute_gamlss.R
#' Multiple Imputation with Generalized Additive Models for Location,
#' Scale, and Shape.
#' 
#' @description Imputes univariate missing data using a generalized
#'   model for location, scale and shape.
#'
#' @param y Numeric vector with incomplete data.
#' @param ry Response pattern of 'y' ('TRUE'=observed,
#'   'FALSE'=missing).
#' @param x Design matrix with 'length(y)' rows and 'p' columns
#'   containing complete covariates.
#' @param family Distribution family to be used by GAMLSS. It defaults
#'   to NO but a range of families can be defined by calling the
#'   corresponding "gamlssFAMILY" method.
#' @param n.ind.par Number of parameters from the distribution family
#'   to be individually estimated.
#' @param gam.mod list with the parameters of the GAMLSS imputation
#'   model.
#' @param fitted.gam A predefined bootstrap gamlss method returned by
#'   \code{fit.gamlss}. Mice by default refits the model with each
#'   imputation. The parameter is here for a future faster modified
#'   mice function.
#' @param EV Logical value to determine whether to correct or not
#'   extreme imputed values. This can arise due to too much
#'   flexibility of the gamlss model.
#' @param ... extra arguments for the control of the gamlss fitting
#'   function
#'
#' @return Numeric vector with imputed values for missing \code{y}
#'   values
#'
#' @details Imputation of \code{y} using generalized additive models
#'   for location, scale, and shape. A model is fitted with the
#'   observed part of the data set. Then a bootstrap sample is
#'   generated and used to refit the model and generate imputations.
#'
#'   The function \code{fit.gamlss} handles the fitting and the
#'   bootstrap and returns a method to generated imputations.
#'
#'   Being gamlss a flexible non parametric method, there may be
#'   problems with the fitting and imputation depending on the sample
#'   size. The imputation functions try to handle anomalies
#'   automatically, but results should be still inspected.
#'
#' @references de Jong, R., van Buuren, S. & Spiess, M. (2016)
#'   Multiple Imputation of Predictor Variables Using Generalized
#'   Additive Models. Communications in Statistics -- Simulation and
#'   Computation, 45(3), 968--985.
#'
#' @references de Jong, Roel. (2012). “Robust Multiple Imputation.” Universität
#' Hamburg. \url{http://ediss.sub.uni-hamburg.de/volltexte/2012/5971/}.
#'
#' @references Rigby, R. A., and Stasinopoulos,
#'   D. M. (2005). Generalized Additive Models for Location, Scale and
#'   Shape. Journal of the Royal Statistical Society: Series C
#'   (Applied Statistics) 54 (3): 507–54.
#'
#' @author Daniel Salfran \email{daniel.salfran@uni-hamburg.de}
#'
#' @examples
#'
#' require(lattice)
#' # Create the imputed data sets
#'
#' predMat <- matrix(rep(0,25), ncol = 5)
#' predMat[4,1] <- 1
#' predMat[4,5] <- 1
#' predMat[2,1] <- 1
#' predMat[2,5] <- 1
#' predMat[2,4] <- 1
#' predMat[3,1] <- 1
#' predMat[3,5] <- 1
#' predMat[3,4] <- 1
#' predMat[3,2] <- 1
#' imputed.sets <- mice(sample.data, m = 2,
#'                      method = c("", "gamlssPO",
#'                                 "gamlss", "gamlssBI", ""),
#'                      visitSequence = "monotone",
#'                      predictorMatrix = predMat,
#'                      maxit = 1, seed = 973,
#'                      n.cyc = 1, bf.cyc = 1,
#'                      cyc = 1)
#'
#' fit <- with(imputed.sets, lm(y ~ X.1 + X.2 + X.3 + X.4))
#' summary(pool(fit))
#'
#' stripplot(imputed.sets)
#'
#' @export
mice.impute.gamlss <- function(y, ry, x, family = NO, n.ind.par = 2,
                               fitted.gam = NULL, gam.mod = list(type = "pb"),
                               EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)

  if (is.null(fitted.gam)) fitted.gam <- do.call(fit.gamlss, as.list(Call)[-1])

  ## imputed.values <- do.call(fitted.gam, as.list(Call)[-1])
  imputed.values <- fitted.gam(...)
  # Repeat the bootstrap step if there is a problem with the fitting
  # of gamlss
  if (sum(is.na(imputed.values)) > 0) {
    imputed.values <- fitted.gam(...)
  }
  if (sum(is.na(imputed.values)) > 0) {
    imputed.values <- fitted.gam(...)
  }
  if (sum(is.na(imputed.values)) > 0) {
    imputed.values <- fitted.gam(...)
  }

  if (EV) {
    if (!(sum(is.na(imputed.values)) == length(imputed.values))) {
      outliers <- getOutliers(imputed.values, rho = c(.3, .3), FLim = c(0.15, 0.85))
      nans <- which(is.nan(imputed.values))
      idx <- c(outliers$iLeft, outliers$iRight, nans)
      if (length(idx) != 0) {
        y[!ry] <- imputed.values
        R = !is.na(y)
        R[!ry][idx] = FALSE
        new.values <- mice.impute.midastouch(y=y, ry=R, x=x)
        imputed.values[idx] <- new.values
      }
    }
  }

  return(imputed.values)
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssNO <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- NO

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssBI <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- BI
  Call[["n.ind.par"]] <- 1

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssJSU <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- JSU

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssPO <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- PO

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssTF <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- TF

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssGA <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- GA

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssZIBI <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- ZIBI

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
mice.impute.gamlssZIP <- function(y, ry, x, fitted.gam = NULL, EV = TRUE, ...) {
  Call <- match.call(expand.dots = TRUE)
  Call[["family"]] <- ZIP

  return(do.call(mice.impute.gamlss, as.list(Call)[-1]))
}

#' @rdname mice.impute.gamlss
#' @export
fit.gamlss <- function(y, ry, x, family = NO, n.ind.par = 2,
                       gam.mod = list(type = "pb"), ...) {

  data <- data.frame(y, x)

  fit <- partial(ImpGamlssFit, family = family, n.ind.par = n.ind.par,
                 gam.mod = gam.mod)

  imp.method <- partial(ImpGamlssBootstrap, fit = fit, R = ry)
  f <- imp.method(incomplete.data = data, ...)

  return(f)
}
